### DISCUSSION TITLE
macOS support?

### DISCUSSION THREAD

**Discussion by czei** [adafruit/Adafruit_Blinka#825] [General]
I'm looking for a way to write a proper test suite for my CircuitPython code that doesn't have to run on a low-level board.  Much of the code I'm writing is pretty low-level, like Bitmap manipulation, but I can only get it to run on low-level hardware without debugger support.   I come from the Java & C++ coding worlds, and abandoning test-driven development to do a Python project is a bitter pill.

This library was recommended to me on the Adafruit forums, but after spending several days trying to get it working on macOS am beginning to wonder if this was the intended purpose at all.   Can it run at all on macOS?   I'm not trying to simulate any hardware at all, just classes inside CircuitPython like Bitmap.  

Thanks,

**Comment by makermelissa** [comment]
It can for certain boards like the FT232H and MCP2221a, but for most boards like the raspberry pi, it is intended to be run on the single board computer.

> I'm not trying to simulate any hardware at all, just classes inside CircuitPython like Bitmap.

Blinka is intended to add a compatibility layer to run CircuitPython code on top of Python. It's possible Blinka might not be the right solution for you. What are you trying to do with the Bitmap? If all you want to do is manipulate it or display it, you might want to look into libraries like Pillow and PyGame.

**Comment by czei** [comment]
Thanks for the suggestion! Those projects look interesting, but quickly perusing their documentation at a high level it doesn't look like they'd run on Adafruit hardware like the MatrixPortal S3.   I could be wrong, but my understanding is CircuitPython wouldn't support all of the language functionality required by libraries designed for the full python library.  

My immediate goal is to manipulate an RGB Display on a MatrixPortal S3 in real time.   

A standard way I approach projects using a small board like this is to:

1. Develop using test-driven methodology on a modern development platform with test suites, a debugger, etc. 
2. Abstract as much logic as possible away from the hardware.

So far I've been unable to develop many tests because few of the CircuitPython libraries will run outside of a board :-(. 

Thanks for any suggestions,

--Michael

**Reply by dhalbert** [reply]
Blinka is one way of simulating some of the native modules that are in CircuitPython. We have also just started working on a `renode` port. `renode` is a simulator. See https://github.com/adafruit/circuitpython/pull/9199 and https://renode.io/. But it is early days.

**Reply by czei** [reply]
Thanks, Dan; renode looks to be precisely what I'm looking for if it is too early of a state to actually use right now.   Note that I'm not expecting any of the hardware interfaces or displays to work--  the low-level classes just need to load to run basic tests against the base classes.   

There are two compatibility questions:

1.  What hardware can Blinka simulate?
2. What platforms does Blinka itself run on?

I tried Blinka on macOS but couldn't get displayio to load when attempting to simulate any hardware at all, which is why I asked the question here about where Blinka can run.   Maybe Blinka only fully runs on Raspberry Pis?   

I believe the answer to #1 from above is that Blinka supports the FT232H and MCP2221a on macOS. And for full support of the current list, I'd have to switch to using the Raspberry Pi as a development platform? Since I'm just trying to load low-level classes, I don't care what hardware it emulates as long as the Bitmap class loads.  Since that wasn't working, I wanted to make sure macOS was on the list of supported platforms before spending more days on something that wasn't intended to work.

As a former software engineer at Sun Microsystems I'm certainly familiar with UNIX/Linux software development, and so *could* switch my development environment to a Raspberry Pi but don't really have time to experiment with something that may or may not work to get a product out the door.  

It strikes me that Adafruit/Circuitpython has the same testing challenge as Sun Microsystem and Solaris:  trying to support code running on various hardware platforms.    I was on the Solaris build team for a while, and there was a room filled with every piece of hardware Sun ever shipped in its history.   There was a huge automated test suite that built Solaris for each platform, installed it, and then ran thousands of tests on each and every piece of hardware.  

The build team was huge at Sun and it was still a lot of work to make sure each release worked on each machine, so I can't imagine the challenge Adafruit has to support so many very different boards with only one or two people.  

Thanks for taking the time to respond,
